tests/filesystem.go

```diff
@@ -0,0 +1,517 @@
+	}
+	return fmt.Sprintf("/dev/mapper/%sp%d", string(match[1]), number)
+}
+
+func writeIgnitionConfig(t *testing.T, config string) string {
+	tmpDir, err := ioutil.TempDir(os.Getenv("TMPDIR"), "config")
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = ioutil.WriteFile(
+		filepath.Join(tmpDir, "config.ign"), []byte(config), 0644)
+	if err != nil {
+		t.Fatal(err)
+	}
+	return tmpDir
+}
+
+func calculateImageSize(partitions []*types.Partition) int64 {
+	// 63 is the number of sectors cgpt uses when generating a hybrid MBR
+	size := int64(63 * 512)
+	for _, p := range partitions {
+		size += int64(align(p.Length, 512) * 512)
+	}
+	size = size + int64(4096*512) // extra room to allow for alignments
+	return size
+}
+
+// createVolume will create the image file of the specified size, create a
+// partition table in it, and generate mount paths for every partition
+func createVolume(t *testing.T, imageFile string, size int64, cylinders int, heads int, sectorsPerTrack int, partitions []*types.Partition) {
+	// attempt to create the file, will leave already existing files alone.
+	// os.Truncate requires the file to already exist
+	out, err := os.Create(imageFile)
+	if err != nil {
+		t.Fatal("create", err, out)
+	}
+	out.Close()
+
+	// Truncate the file to the given size
+	err = os.Truncate(imageFile, size)
+	if err != nil {
+		t.Fatal("truncate", err)
+	}
+
+	createPartitionTable(t, imageFile, partitions)
+
+	for counter, partition := range partitions {
+		if partition.TypeCode == "blank" || partition.FilesystemType == "" {
+			continue
+		}
+
+		mntPath, err := ioutil.TempDir(os.Getenv("TMPDIR"), fmt.Sprintf("hd1p%d", counter))
+		if err != nil {
+			t.Fatal(err)
+		}
+		partition.MountPath = mntPath
+	}
+}
+
+// setDevices will create devices for each of the partitions in the imageFile,
+// and then will format each partition according to what's descrived in the
+// partitions argument.
+func setDevices(t *testing.T, imageFile string, partitions []*types.Partition) string {
+	loopDevice := kpartxAdd(t, imageFile)
+
+	for _, partition := range partitions {
+		if partition.TypeCode == "blank" || partition.FilesystemType == "" {
+			continue
+		}
+
+		partition.Device = fmt.Sprintf(
+			"/dev/mapper/%sp%d", loopDevice, partition.Number)
+		formatPartition(t, partition)
+	}
+	return fmt.Sprintf("/dev/%s", loopDevice)
+}
+
+func destroyDevices(t *testing.T, imageFile string) {
+	args := []string{"-d", imageFile}
+	_ = mustRun(t, "kpartx", args...)
+}
+func formatPartition(t *testing.T, partition *types.Partition) {
+	var mkfs string
+	var opts, label, uuid []string
+
+	switch partition.FilesystemType {
+	case "vfat":
+		mkfs = "mkfs.vfat"
+		if partition.FilesystemLabel != "" {
+			label = []string{"-n", partition.FilesystemLabel}
+		}
+		if partition.FilesystemUUID != "" {
+			uuid = []string{"-i", partition.FilesystemUUID}
+		}
+	case "ext2", "ext4":
+		mkfs = "mke2fs"
+		opts = []string{
+			"-t", partition.FilesystemType, "-b", "4096",
```

Comment:
As previously mentioned, this and the tune2fs command need a comment saying that they're mirroring `disk_util`.